src/Cli/Microsoft.DotNet.Cli.Utils/Reporter.cs

```diff
@@ -23,32 +32,87 @@ private Reporter(AnsiConsole? console)
+
+        public static IReporter Output { get; private set; } = NullReporter;
+        public static IReporter Error { get; private set; } = NullReporter;
+        public static IReporter Verbose { get; private set; } = NullReporter;
 
         /// <summary>
-        /// Resets the Reporters to write to the current Console Out/Error.
+        /// Resets the reporters to write to the current reporters based on <see cref="CommandLoggingContext"/> settings.
         /// </summary>
         public static void Reset()
         {
-            lock (_lock)
+            lock (s_lock)
             {
-                Output = new Reporter(AnsiConsole.GetOutput());
-                Error = new Reporter(AnsiConsole.GetError());
-                Verbose = IsVerbose ?
-                    new Reporter(AnsiConsole.GetOutput()) :
-                    NullReporter;
+                ResetOutput();
+                ResetError();
+                ResetVerbose();
             }
         }
 
-        public static bool IsVerbose => CommandContext.IsVerbose();
+        /// <summary>
+        /// Sets the output reporter to <paramref name="reporter"/>.
+        /// The reporter won't be applied if disabled in <see cref="CommandLoggingContext"/>.
+        /// </summary>
+        /// <param name="reporter"></param>
+        public static void SetOutput(IReporter reporter)
+        {
+            lock (s_lock)
+            {
+                s_outputReporter = reporter;
+                ResetOutput();
+            }
+        }
+
+        /// <summary>
+        /// Sets the error reporter to <paramref name="reporter"/>.
+        /// The reporter won't be applied if disabled in <see cref="CommandLoggingContext"/>.
+        /// </summary>
+        public static void SetError(IReporter reporter)
+        {
+            lock (s_lock)
+            {
+                s_errorReporter = reporter;
+                ResetError();
+            }
+        }
+
+        /// <summary>
+        /// Sets the verbose reporter to <paramref name="reporter"/>.
+        /// The reporter won't be applied if disabled in <see cref="CommandLoggingContext"/>.
+        /// </summary>
+        public static void SetVerbose(IReporter reporter)
+        {
+            lock (s_lock)
+            {
+                s_verboseReporter = reporter;
+                ResetVerbose();
+            }
+        }
+
+        private static void ResetOutput()
+        {
+            Output = CommandLoggingContext.OutputEnabled ? s_outputReporter : NullReporter;
+        }
+
+        private static void ResetError()
+        {
+            Error = CommandLoggingContext.ErrorEnabled ? s_errorReporter : NullReporter;
+        }
+
+        private static void ResetVerbose()
+        {
+            Verbose = CommandLoggingContext.IsVerbose ? s_verboseReporter : NullReporter;
+        }
 
         public void WriteLine(string message)
         {
-            lock (_lock)
+            lock (s_lock)
```

Comment:
Since we're touching this - I'd replace lock with `SpinLock` or other more lightweight synchronization.
We'll be seldomly altering the state and often just accessing it.

Btw. is it really needed to lock even the console writes? This feels to heavy

Reply:
This is very old code. I may assume that is done so 2 writes does not happen at the time changing text/colors/etc.
@dsplaisted @joeloff do you know any known issues that actually requires this synchronization?

Reply:
No, no idea.

Reply:
Switched to `SpinLock`: if there are any objects please let me know asap.